include/votca/csg/boundarycondition.h

```diff
@@ -15,52 +15,81 @@
  *
  */
 
-#ifndef _VOTCA_CSG_BOUNDARYCONDITION_H
-#define _VOTCA_CSG_BOUNDARYCONDITION_H
+#ifndef VOTCA_CSG_BOUNDARYCONDITION_H
+#define VOTCA_CSG_BOUNDARYCONDITION_H
 
-#include <votca/tools/matrix.h>
+#include <memory>
+#include <stdexcept>
+#include <utility>
+#include <votca/tools/eigen.h>
 
 namespace votca {
 namespace csg {
-using namespace votca::tools;
 
+/**
+ * @brief Class keeps track of how the boundaries of the system are handled
+ *
+ * There are a total of 3 different boundaries:
+ * open - no boundaries
+ * orthorhombic - orthorombic boundaries
+ * triclinic - triclinic boundaries
+ *
+ * This class enables the correct treatement of distances beteween topology
+ * objects, such that distances accound for the periodic boundaries.
+ */
 class BoundaryCondition {
 
  public:
   virtual ~BoundaryCondition(){};
 
+  /**
+   * @brief Safe way to allow child classes to be copied
+   *
+   * The child classes must use the same method and override it with their type
+   * for this to work.
+   *
+   * @return standard pointer to child class
+   */
+  virtual std::unique_ptr<BoundaryCondition> Clone() const = 0;
   /**
    * set the simulation box
    * \param box triclinic box matrix
    */
-  void setBox(const matrix &box) { _box = box; };
+  void setBox(const Eigen::Matrix3d &box) { box_ = box; };
 
   /**
    * get the simulation box
    * \return triclinic box matrix
    */
-  const matrix &getBox() { return _box; };
+  const Eigen::Matrix3d &getBox() { return box_; };
 
   /**
    * get the volume of the box
    * \return box volume as double
    */
-  virtual double BoxVolume();
+  virtual double BoxVolume() const;
 
   /**
    * get shortest connection vector between r_i and r_j with respect to the
    * (periodic) box \return shortest distance vector
    */
-  virtual vec BCShortestConnection(const vec &r_i, const vec &r_j) const = 0;
+  virtual Eigen::Vector3d BCShortestConnection(
+      const Eigen::Vector3d &r_i, const Eigen::Vector3d &r_j) const {
```

Comment:
no make it a pure virtual function

Reply:
you are creating a runtime error, where before you got a compiler error. That is not good